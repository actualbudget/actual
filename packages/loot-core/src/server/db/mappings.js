import { addSyncListener } from '../sync/index';

import * as db from './index';

// This file keeps all the mappings in memory so we can access it
// synchronously. This is primarily used in the rules system, but
// there may be other uses in the future. You don't need to worry
// about this generally; if you are querying transactions, ids are
// transparently mapped for you. But if you are building something
// that stores ids and later uses them, you need to remember to map
// the ids.
//
// IMPORTANT: `loadMappings` must be called first before other modules
// that listen for sync changes. This must be the first sync listener
// to run in case other listeners use this mapping table; otherwise
// they might see stale mappings.

let allMappings;
let unlistenSync;

export async function loadMappings() {
  // The mappings are separated into tables specific to the type of
  // data. But you know, we really could keep a global mapping table.
  let categories = (await db.all('SELECT * FROM category_mapping')).map(r => [
    r.id,
    r.transferId,
  ]);
  let payees = (await db.all('SELECT * FROM payee_mapping')).map(r => [
    r.id,
    r.targetId,
  ]);

  // All ids are unique, so we can just keep a global table of mappings
  allMappings = new Map(categories.concat(payees));

  if (unlistenSync) {
    unlistenSync();
  }
  unlistenSync = addSyncListener(onApplySync);
}

function onApplySync(oldValues, newValues) {
  newValues.forEach((items, table) => {
    if (table.indexOf('mapping') !== -1) {
      let field = table === 'category_mapping' ? 'transferId' : 'targetId';

      items.forEach(newValue => {
        allMappings.set(newValue.id, newValue[field]);
      });
    }
  });
}

export function getMappings() {
  return allMappings;
}

export function getMapping(id) {
  return allMappings.get(id) || null;
}
