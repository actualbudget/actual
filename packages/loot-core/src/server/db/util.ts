export async function incrFetch(
  runQuery,
  terms,
  compare,
  makeQuery,
  params = [],
) {
  let pageCount = 500;
  let results = [];

  let fetchedIds = new Set();

  for (let i = 0; i < terms.length; i += pageCount) {
    let slice = terms.slice(i, i + pageCount).filter(id => !fetchedIds.has(id));
    if (slice.length > 0) {
      let filter = slice.map(id => compare(id)).join(' OR ');
      let query = makeQuery('(' + filter + ')');

      let rows = await runQuery(query, params, true);
      fetchedIds = new Set([...fetchedIds, ...slice]);
      results = results.concat(rows);
    }
  }

  return results;
}

export function whereIn(ids: string[], field: string) {
  let ids2 = [...new Set(ids)];
  // eslint-disable-next-line rulesdir/typography
  let filter = `${field} IN (` + ids2.map(id => `'${id}'`).join(',') + ')';
  return filter;
}
