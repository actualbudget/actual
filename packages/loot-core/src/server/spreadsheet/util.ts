export function unresolveName(name) {
  let idx = name.indexOf('!');
  if (idx !== -1) {
    return {
      sheet: name.slice(0, idx),
      name: name.slice(idx + 1),
    };
  }
  return { sheet: null, name };
}

export function resolveName(sheet: string, name: string): string {
  return sheet + '!' + name;
}
