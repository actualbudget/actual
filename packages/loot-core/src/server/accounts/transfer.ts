import * as db from '../db';

async function getPayee(acct) {
  return db.first('SELECT * FROM payees WHERE transfer_acct = ?', [acct]);
}

async function getTransferredAccount(transaction) {
  if (transaction.payee) {
    let { transfer_acct } = await db.first(
      'SELECT id, transfer_acct FROM v_payees WHERE id = ?',
      [transaction.payee],
    );
    return transfer_acct;
  }
  return null;
}

async function clearCategory(transaction, transferAcct) {
  const { offbudget: fromOffBudget } = await db.first(
    'SELECT offbudget FROM accounts WHERE id = ?',
    [transaction.account],
  );
  const { offbudget: toOffBudget } = await db.first(
    'SELECT offbudget FROM accounts WHERE id = ?',
    [transferAcct],
  );

  // We should clear the category to make sure it's not being
  // accounted for in the budget, unless it should be in the case of
  // transferring from an on-budget to off-budget account
  if (fromOffBudget === toOffBudget) {
    await db.updateTransaction({ id: transaction.id, category: null });
    return true;
  }
  return false;
}

export async function addTransfer(transaction, transferredAccount) {
  if (transaction.is_parent) {
    // For split transactions, we should create transfers using child transactions.
    // This is to ensure that the amounts received by the transferred account
    // reflects the amounts in the child transactions and not the parent transaction
    // amount which is the total amount.
    return null;
  }

  let { id: fromPayee } = await db.first(
    'SELECT id FROM payees WHERE transfer_acct = ?',
    [transaction.account],
  );

  // We need to enforce certain constraints with child transaction transfers
  if (transaction.parent_id) {
    let row = await db.first(
      `
        SELECT p.id, p.transfer_acct FROM v_transactions t
        LEFT JOIN payees p ON p.id = t.payee
        WHERE t.id = ?
      `,
      [transaction.parent_id],
    );

    if (row.transfer_acct) {
      if (row.id !== transaction.payee) {
        // This child transaction is trying to use a transfer payee,
        // but the parent is already using a different transfer payee.
        // This is not allowed, so not only do we do nothing, we clear
        // the payee of the child transaction to make it clear
        await db.updateTransaction({ id: transaction.id, payee: null });
        return { id: transaction.id, payee: null };
      }
    }
  }

  const id = await db.insertTransaction({
    account: transferredAccount,
    amount: -transaction.amount,
    payee: fromPayee,
    date: transaction.date,
    transfer_id: transaction.id,
    notes: transaction.notes || null,
    cleared: false,
  });

  await db.updateTransaction({ id: transaction.id, transfer_id: id });
  const categoryCleared = await clearCategory(transaction, transferredAccount);

  return {
    id: transaction.id,
    transfer_id: id,
    ...(categoryCleared ? { category: null } : {}),
  };
}

export async function removeTransfer(transaction) {
  let transferTrans = await db.getTransaction(transaction.transfer_id);

  // Perform operations on the transfer transaction only
  // if it is found. For example: when users delete both
  // (in & out) transfer transactions at the same time -
  // transfer transaction will not be found.
  if (transferTrans) {
    if (transferTrans.is_child) {
      // If it's a child transaction, we don't delete it because that
      // would invalidate the whole split transaction. Instead of turn
      // it into a normal transaction
      await db.updateTransaction({
        id: transaction.transfer_id,
        transfer_id: null,
        payee: null,
      });
    } else {
      await db.deleteTransaction({ id: transaction.transfer_id });
    }
  }
  await db.updateTransaction({ id: transaction.id, transfer_id: null });
  return { id: transaction.id, transfer_id: null };
}

export async function updateTransfer(transaction, transferredAccount) {
  let payee = await getPayee(transaction.account);

  await db.updateTransaction({
    id: transaction.transfer_id,
    account: transferredAccount,
    // Make sure to update the payee on the other side in case the
    // user moved this transaction into another account
    payee: payee.id,
    date: transaction.date,
    notes: transaction.notes,
    amount: -transaction.amount,
  });

  const categoryCleared = await clearCategory(transaction, transferredAccount);
  if (categoryCleared) {
    return { id: transaction.id, category: null };
  }
}

export async function onInsert(transaction) {
  let transferredAccount = await getTransferredAccount(transaction);

  if (transferredAccount) {
    return addTransfer(transaction, transferredAccount);
  }
}

export async function onDelete(transaction) {
  if (transaction.transfer_id) {
    await removeTransfer(transaction);
  }
}

export async function onUpdate(transaction) {
  const transferredAccount = await getTransferredAccount(transaction);

  if (transferredAccount && !transaction.transfer_id) {
    return addTransfer(transaction, transferredAccount);
  }

  if (!transferredAccount && transaction.transfer_id) {
    return removeTransfer(transaction);
  }

  if (transferredAccount && transaction.transfer_id) {
    return updateTransfer(transaction, transferredAccount);
  }
}
