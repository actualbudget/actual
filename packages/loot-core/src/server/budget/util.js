import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return amounts.reduce((total, amount) => {
    return total + number(amount);
  }, 0);
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}

// Note that we don't restrict values to `Number.MIN_SAFE_INTEGER <= value <= Number.MAX_SAFE_INTEGER`
// where `Number.MAX_SAFE_INTEGER == 2^53 - 1` but a smaller range over `-(2^43-1) <= value <= 2^43 - 1`.
// This ensure that the number is accurate not just for the integer component but for 3 decimal places also.
//
// This gives us the guarantee that can use `safeNumber` on number whether they are unscaled user inputs
// or they have been converted to integers (using `amountToInteger`).

const MAX_SAFE_NUMBER = 2 ** 43 - 1;
const MIN_SAFE_NUMBER = -MAX_SAFE_NUMBER;

export function safeNumber(value) {
  value = number(value);
  if (value > MAX_SAFE_NUMBER || value < MIN_SAFE_NUMBER) {
    throw new Error(
      "Can't safely perform arithmetic with number: " + JSON.stringify(value)
    );
  }
  return value;
}
