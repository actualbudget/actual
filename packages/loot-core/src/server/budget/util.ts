import { safeNumber } from '../../shared/util';
import { number } from '../spreadsheet/globals';

export { number } from '../spreadsheet/globals';

export function sumAmounts(...amounts) {
  return safeNumber(
    amounts.reduce((total, amount) => {
      return total + number(amount);
    }, 0),
  );
}

export function flatten2(arr) {
  return Array.prototype.concat.apply([], arr);
}

export function unflatten2(arr) {
  let res = [];
  for (let i = 0; i < arr.length; i += 2) {
    res.push([arr[i], arr[i + 1]]);
  }
  return res;
}
