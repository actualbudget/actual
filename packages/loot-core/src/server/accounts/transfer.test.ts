import { expectSnapshotWithDiffer } from '../../mocks/util';
import * as db from '../db';

import * as transfer from './transfer';

beforeEach(global.emptyDatabase());

function getAllTransactions() {
  return db.all(
    `SELECT t.*, p.name as payee_name
       FROM v_transactions t
       LEFT JOIN payees p ON p.id = t.payee
       ORDER BY date DESC, amount DESC, id
     `,
  );
}

async function prepareDatabase() {
  await db.insertCategoryGroup({ id: 'group1', name: 'group1', is_income: 0 });
  await db.insertCategory({
    id: '1',
    name: 'cat1',
    cat_group: 'group1',
    is_income: 0,
  });
  await db.insertAccount({ id: 'one', name: 'one' });
  await db.insertAccount({ id: 'two', name: 'two' });
  await db.insertAccount({ id: 'three', name: 'three', offbudget: 1 });
  await db.insertPayee({ name: '', transfer_acct: 'one' });
  await db.insertPayee({ name: '', transfer_acct: 'two' });
  await db.insertPayee({
    name: '',
    transfer_acct: 'three',
  });
}

type Transaction = {
  account: string;
  amount: number;
  category?: string;
  date: string;
  id?: string;
  notes?: string;
  payee: string;
  transfer_id?: string;
};

describe('Transfer', () => {
  test('transfers are properly inserted/updated/deleted', async () => {
    await prepareDatabase();

    let transaction: Transaction = {
      account: 'one',
      amount: 5000,
      payee: await db.insertPayee({ name: 'Non-transfer' }),
      date: '2017-01-01',
    };
    await db.insertTransaction(transaction);
    await transfer.onInsert(transaction);

    let differ = expectSnapshotWithDiffer(await getAllTransactions());

    let transferTwo = await db.first(
      "SELECT * FROM payees WHERE transfer_acct = 'two'",
    );
    let transferThree = await db.first(
      "SELECT * FROM payees WHERE transfer_acct = 'three'",
    );

    transaction = {
      account: 'one',
      amount: 5000,
      payee: transferTwo.id,
      date: '2017-01-01',
    };
    transaction.id = await db.insertTransaction(transaction);
    await transfer.onInsert(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    // Fill the transaction out
    transaction = await db.getTransaction(transaction.id);
    expect(transaction.transfer_id).toBeDefined();

    transaction = {
      ...transaction,
      date: '2017-01-05',
      notes: 'This is a note',
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    transaction = {
      ...transaction,
      payee: transferThree.id,
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    transaction = {
      ...transaction,
      payee: await db.insertPayee({ name: 'Not transferred anymore' }),
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    // Make sure it's not a linked transaction anymore
    transaction = await db.getTransaction(transaction.id);
    expect(transaction.transfer_id).toBeNull();

    // Re-transfer it
    transaction = {
      ...transaction,
      payee: transferTwo.id,
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    transaction = await db.getTransaction(transaction.id);
    expect(transaction.transfer_id).toBeDefined();

    await db.deleteTransaction(transaction);
    await transfer.onDelete(transaction);
    differ.expectToMatchDiff(await getAllTransactions());
  });

  test('transfers are properly de-categorized', async () => {
    await prepareDatabase();

    let transferTwo = await db.first(
      "SELECT * FROM payees WHERE transfer_acct = 'two'",
    );
    let transferThree = await db.first(
      "SELECT * FROM payees WHERE transfer_acct = 'three'",
    );

    let transaction: Transaction = {
      account: 'one',
      amount: 5000,
      payee: await db.insertPayee({ name: 'Non-transfer' }),
      date: '2017-01-01',
      category: '1',
    };
    transaction.id = await db.insertTransaction(transaction);
    await transfer.onInsert(transaction);

    let differ = expectSnapshotWithDiffer(await getAllTransactions());

    transaction = {
      ...(await db.getTransaction(transaction.id)),
      payee: transferThree.id,
      notes: 'hi',
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());

    transaction = {
      ...(await db.getTransaction(transaction.id)),
      payee: transferTwo.id,
    };
    await db.updateTransaction(transaction);
    await transfer.onUpdate(transaction);
    differ.expectToMatchDiff(await getAllTransactions());
  });
});
