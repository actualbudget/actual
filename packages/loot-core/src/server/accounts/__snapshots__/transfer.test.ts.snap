// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transfer transfers are properly de-categorized 1`] = `
Array [
  Object {
    "account": "one",
    "amount": 5000,
    "category": "1",
    "cleared": 1,
    "date": 20170101,
    "error": null,
    "id": "id6",
    "imported_id": null,
    "imported_payee": null,
    "is_child": 0,
    "is_parent": 0,
    "notes": null,
    "parent_id": null,
    "payee": "id5",
    "payee_name": "Non-transfer",
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transfer_id": null,
  },
]
`;

exports[`Transfer transfers are properly de-categorized 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -9,16 +9,38 @@
      \\"id\\": \\"id6\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
-     \\"notes\\": null,
+     \\"notes\\": \\"hi\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id5\\",
-     \\"payee_name\\": \\"Non-transfer\\",
+     \\"payee\\": \\"id4\\",
+     \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": null,
+     \\"transfer_id\\": \\"id7\\",
+   },
+   Object {
+     \\"account\\": \\"three\\",
+     \\"amount\\": -5000,
+     \\"category\\": null,
+     \\"cleared\\": 0,
+     \\"date\\": 20170101,
+     \\"error\\": null,
+     \\"id\\": \\"id7\\",
+     \\"imported_id\\": null,
+     \\"imported_payee\\": null,
+     \\"is_child\\": 0,
+     \\"is_parent\\": 0,
+     \\"notes\\": \\"hi\\",
+     \\"parent_id\\": null,
+     \\"payee\\": \\"id2\\",
+     \\"payee_name\\": \\"\\",
+     \\"schedule\\": null,
+     \\"sort_order\\": 123456789,
+     \\"starting_balance_flag\\": 0,
+     \\"tombstone\\": 0,
+     \\"transfer_id\\": \\"id6\\",
    },
  ]"
`;

exports[`Transfer transfers are properly de-categorized 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -1,30 +1,30 @@
  Array [
    Object {
      \\"account\\": \\"one\\",
      \\"amount\\": 5000,
-     \\"category\\": \\"1\\",
+     \\"category\\": null,
      \\"cleared\\": 1,
      \\"date\\": 20170101,
      \\"error\\": null,
      \\"id\\": \\"id6\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
      \\"notes\\": \\"hi\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id4\\",
+     \\"payee\\": \\"id3\\",
      \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
      \\"transfer_id\\": \\"id7\\",
    },
    Object {
-     \\"account\\": \\"three\\",
+     \\"account\\": \\"two\\",
      \\"amount\\": -5000,
      \\"category\\": null,
      \\"cleared\\": 0,
      \\"date\\": 20170101,
      \\"error\\": null,"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 1`] = `
Array [
  Object {
    "account": "one",
    "amount": 5000,
    "category": null,
    "cleared": 1,
    "date": 20170101,
    "error": null,
    "id": "id6",
    "imported_id": null,
    "imported_payee": null,
    "is_child": 0,
    "is_parent": 0,
    "notes": null,
    "parent_id": null,
    "payee": "id5",
    "payee_name": "Non-transfer",
    "schedule": null,
    "sort_order": 123456789,
    "starting_balance_flag": 0,
    "tombstone": 0,
    "transfer_id": null,
  },
]
`;

exports[`Transfer transfers are properly inserted/updated/deleted 2`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -19,6 +19,50 @@
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
      \\"transfer_id\\": null,
    },
+   Object {
+     \\"account\\": \\"one\\",
+     \\"amount\\": 5000,
+     \\"category\\": null,
+     \\"cleared\\": 1,
+     \\"date\\": 20170101,
+     \\"error\\": null,
+     \\"id\\": \\"id7\\",
+     \\"imported_id\\": null,
+     \\"imported_payee\\": null,
+     \\"is_child\\": 0,
+     \\"is_parent\\": 0,
+     \\"notes\\": null,
+     \\"parent_id\\": null,
+     \\"payee\\": \\"id3\\",
+     \\"payee_name\\": \\"\\",
+     \\"schedule\\": null,
+     \\"sort_order\\": 123456789,
+     \\"starting_balance_flag\\": 0,
+     \\"tombstone\\": 0,
+     \\"transfer_id\\": \\"id8\\",
+   },
+   Object {
+     \\"account\\": \\"two\\",
+     \\"amount\\": -5000,
+     \\"category\\": null,
+     \\"cleared\\": 0,
+     \\"date\\": 20170101,
+     \\"error\\": null,
+     \\"id\\": \\"id8\\",
+     \\"imported_id\\": null,
+     \\"imported_payee\\": null,
+     \\"is_child\\": 0,
+     \\"is_parent\\": 0,
+     \\"notes\\": null,
+     \\"parent_id\\": null,
+     \\"payee\\": \\"id2\\",
+     \\"payee_name\\": \\"\\",
+     \\"schedule\\": null,
+     \\"sort_order\\": 123456789,
+     \\"starting_balance_flag\\": 0,
+     \\"tombstone\\": 0,
+     \\"transfer_id\\": \\"id7\\",
+   },
  ]"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 3`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,67 +2,67 @@
    Object {
      \\"account\\": \\"one\\",
      \\"amount\\": 5000,
      \\"category\\": null,
      \\"cleared\\": 1,
-     \\"date\\": 20170101,
+     \\"date\\": 20170105,
      \\"error\\": null,
-     \\"id\\": \\"id6\\",
+     \\"id\\": \\"id7\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
-     \\"notes\\": null,
+     \\"notes\\": \\"This is a note\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id5\\",
-     \\"payee_name\\": \\"Non-transfer\\",
+     \\"payee\\": \\"id3\\",
+     \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": null,
+     \\"transfer_id\\": \\"id8\\",
    },
    Object {
-     \\"account\\": \\"one\\",
-     \\"amount\\": 5000,
+     \\"account\\": \\"two\\",
+     \\"amount\\": -5000,
      \\"category\\": null,
-     \\"cleared\\": 1,
-     \\"date\\": 20170101,
+     \\"cleared\\": 0,
+     \\"date\\": 20170105,
      \\"error\\": null,
-     \\"id\\": \\"id7\\",
+     \\"id\\": \\"id8\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
-     \\"notes\\": null,
+     \\"notes\\": \\"This is a note\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id3\\",
+     \\"payee\\": \\"id2\\",
      \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id8\\",
+     \\"transfer_id\\": \\"id7\\",
    },
    Object {
-     \\"account\\": \\"two\\",
-     \\"amount\\": -5000,
+     \\"account\\": \\"one\\",
+     \\"amount\\": 5000,
      \\"category\\": null,
-     \\"cleared\\": 0,
+     \\"cleared\\": 1,
      \\"date\\": 20170101,
      \\"error\\": null,
-     \\"id\\": \\"id8\\",
+     \\"id\\": \\"id6\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
      \\"notes\\": null,
      \\"parent_id\\": null,
-     \\"payee\\": \\"id2\\",
-     \\"payee_name\\": \\"\\",
+     \\"payee\\": \\"id5\\",
+     \\"payee_name\\": \\"Non-transfer\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id7\\",
+     \\"transfer_id\\": null,
    },
  ]"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 4`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -11,20 +11,20 @@
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
      \\"notes\\": \\"This is a note\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id3\\",
+     \\"payee\\": \\"id4\\",
      \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
      \\"transfer_id\\": \\"id8\\",
    },
    Object {
-     \\"account\\": \\"two\\",
+     \\"account\\": \\"three\\",
      \\"amount\\": -5000,
      \\"category\\": null,
      \\"cleared\\": 0,
      \\"date\\": 20170105,
      \\"error\\": null,"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 5`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -11,39 +11,17 @@
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
      \\"notes\\": \\"This is a note\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id4\\",
-     \\"payee_name\\": \\"\\",
+     \\"payee\\": \\"id9\\",
+     \\"payee_name\\": \\"Not transferred anymore\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id8\\",
-   },
-   Object {
-     \\"account\\": \\"three\\",
-     \\"amount\\": -5000,
-     \\"category\\": null,
-     \\"cleared\\": 0,
-     \\"date\\": 20170105,
-     \\"error\\": null,
-     \\"id\\": \\"id8\\",
-     \\"imported_id\\": null,
-     \\"imported_payee\\": null,
-     \\"is_child\\": 0,
-     \\"is_parent\\": 0,
-     \\"notes\\": \\"This is a note\\",
-     \\"parent_id\\": null,
-     \\"payee\\": \\"id2\\",
-     \\"payee_name\\": \\"\\",
-     \\"schedule\\": null,
-     \\"sort_order\\": 123456789,
-     \\"starting_balance_flag\\": 0,
-     \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id7\\",
+     \\"transfer_id\\": null,
    },
    Object {
      \\"account\\": \\"one\\",
      \\"amount\\": 5000,
      \\"category\\": null,"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 6`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -11,17 +11,39 @@
      \\"imported_payee\\": null,
      \\"is_child\\": 0,
      \\"is_parent\\": 0,
      \\"notes\\": \\"This is a note\\",
      \\"parent_id\\": null,
-     \\"payee\\": \\"id9\\",
-     \\"payee_name\\": \\"Not transferred anymore\\",
+     \\"payee\\": \\"id3\\",
+     \\"payee_name\\": \\"\\",
      \\"schedule\\": null,
      \\"sort_order\\": 123456789,
      \\"starting_balance_flag\\": 0,
      \\"tombstone\\": 0,
-     \\"transfer_id\\": null,
+     \\"transfer_id\\": \\"id10\\",
+   },
+   Object {
+     \\"account\\": \\"two\\",
+     \\"amount\\": -5000,
+     \\"category\\": null,
+     \\"cleared\\": 0,
+     \\"date\\": 20170105,
+     \\"error\\": null,
+     \\"id\\": \\"id10\\",
+     \\"imported_id\\": null,
+     \\"imported_payee\\": null,
+     \\"is_child\\": 0,
+     \\"is_parent\\": 0,
+     \\"notes\\": \\"This is a note\\",
+     \\"parent_id\\": null,
+     \\"payee\\": \\"id2\\",
+     \\"payee_name\\": \\"\\",
+     \\"schedule\\": null,
+     \\"sort_order\\": 123456789,
+     \\"starting_balance_flag\\": 0,
+     \\"tombstone\\": 0,
+     \\"transfer_id\\": \\"id7\\",
    },
    Object {
      \\"account\\": \\"one\\",
      \\"amount\\": 5000,
      \\"category\\": null,"
`;

exports[`Transfer transfers are properly inserted/updated/deleted 7`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -2,54 +2,10 @@
    Object {
      \\"account\\": \\"one\\",
      \\"amount\\": 5000,
      \\"category\\": null,
      \\"cleared\\": 1,
-     \\"date\\": 20170105,
-     \\"error\\": null,
-     \\"id\\": \\"id7\\",
-     \\"imported_id\\": null,
-     \\"imported_payee\\": null,
-     \\"is_child\\": 0,
-     \\"is_parent\\": 0,
-     \\"notes\\": \\"This is a note\\",
-     \\"parent_id\\": null,
-     \\"payee\\": \\"id3\\",
-     \\"payee_name\\": \\"\\",
-     \\"schedule\\": null,
-     \\"sort_order\\": 123456789,
-     \\"starting_balance_flag\\": 0,
-     \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id10\\",
-   },
-   Object {
-     \\"account\\": \\"two\\",
-     \\"amount\\": -5000,
-     \\"category\\": null,
-     \\"cleared\\": 0,
-     \\"date\\": 20170105,
-     \\"error\\": null,
-     \\"id\\": \\"id10\\",
-     \\"imported_id\\": null,
-     \\"imported_payee\\": null,
-     \\"is_child\\": 0,
-     \\"is_parent\\": 0,
-     \\"notes\\": \\"This is a note\\",
-     \\"parent_id\\": null,
-     \\"payee\\": \\"id2\\",
-     \\"payee_name\\": \\"\\",
-     \\"schedule\\": null,
-     \\"sort_order\\": 123456789,
-     \\"starting_balance_flag\\": 0,
-     \\"tombstone\\": 0,
-     \\"transfer_id\\": \\"id7\\",
-   },
-   Object {
-     \\"account\\": \\"one\\",
-     \\"amount\\": 5000,
-     \\"category\\": null,
-     \\"cleared\\": 1,
      \\"date\\": 20170101,
      \\"error\\": null,
      \\"id\\": \\"id6\\",
      \\"imported_id\\": null,
      \\"imported_payee\\": null,"
`;
